google/cloud/bigtable/row_merger.py

```diff
@@ -0,0 +1,497 @@
+
+            if self.state == _State.ROW_START:
+                self._handle_row_start(chunk)
+
+            if self.state == _State.CELL_START:
+                self._handle_cell_start(chunk)
+
+            if self.state == _State.CELL_IN_PROGRESS:
+                self._handle_cell_in_progress(chunk)
+
+            if self.state == _State.CELL_COMPLETE:
+                self._handle_cell_complete(chunk)
+
+            if self.state == _State.ROW_COMPLETE:
+                yield self._handle_row_complete(chunk)
+            elif chunk.commit_row:
+                raise InvalidChunk(
+                    f"Chunk tried to commit row in wrong state (${self.state})"
+                )
+
+    def _handle_reset(self, chunk):
+        if self.state == _State.ROW_START:
+            raise InvalidChunk("Bare reset")
+        if chunk.row_key:
+            raise InvalidChunk("Reset chunk has a row key")
+        if chunk.HasField("family_name"):
+            raise InvalidChunk("Reset chunk has family_name")
+        if chunk.HasField("qualifier"):
+            raise InvalidChunk("Reset chunk has qualifier")
+        if chunk.timestamp_micros:
+            raise InvalidChunk("Reset chunk has a timestamp")
+        if chunk.labels:
+            raise InvalidChunk("Reset chunk has labels")
+        if chunk.value:
+            raise InvalidChunk("Reset chunk has a value")
+
+        self.state = _State.ROW_START
+        self.row = None
+
+    def _handle_row_start(self, chunk):
+        if not chunk.row_key:
+            raise InvalidChunk("New row is missing a row key")
+        if self.last_seen_row_key and self.last_seen_row_key >= chunk.row_key:
+            raise InvalidChunk("Out of order row keys")
+
+        self.row = _PartialRow(chunk.row_key)
+        self.state = _State.CELL_START
+
+    def _handle_cell_start(self, chunk):
+        # Ensure that all chunks after the first one either are missing a row
+        # key or the row is the same
+        if self.row.cells and chunk.row_key and chunk.row_key != self.row.row_key:
+            raise InvalidChunk("row key changed mid row")
+
+        if not self.row.cell:
+            self.row.cell = _PartialCell()
+
+        if chunk.HasField("family_name"):
+            self.row.cell.family = chunk.family_name.value
+            self.row.cell.qualifier = None
+        if not self.row.cell.family:
+            raise InvalidChunk("missing family for a new cell")
+
+        if chunk.HasField("qualifier"):
+            self.row.cell.qualifier = chunk.qualifier.value
+        if self.row.cell.qualifier is None:
+            raise InvalidChunk("missing family for a new cell")
+
+        self.row.cell.timestamp = chunk.timestamp_micros
+        self.row.cell.labels = chunk.labels
+
+        if chunk.value_size > 0:
+            # explicitly avoid pre-allocation as it seems that bytearray
+            # concatenation performs better than slice copies.
+            self.row.cell.value = bytearray()
```

Comment:
how will we add the current chunk to the cell value? should we add self.row.cell.value += chunk.value to line 406?